pkg/azuredx/datasource.go

```diff
@@ -1,30 +1,62 @@
-package main
+package azuredx
 
 import (
 	"fmt"
+	"log"
 
-	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx"
+	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx/client"
+	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx/models"
+	"github.com/grafana/azure-data-explorer-datasource/pkg/azuredx/tokenprovider"
 	"github.com/grafana/grafana-plugin-sdk-go/backend"
+	"github.com/grafana/grafana-plugin-sdk-go/backend/instancemgmt"
 	"github.com/grafana/grafana-plugin-sdk-go/data"
 	jsoniter "github.com/json-iterator/go"
 	"golang.org/x/net/context"
 )
 
 // GrafanaAzureDXDatasource stores reference to plugin and logger
-type GrafanaAzureDXDatasource struct{}
+type GrafanaAzureDXDatasource struct {
+	client   *client.Client
+	settings *models.DatasourceSettings
+}
 
-func (plugin *GrafanaAzureDXDatasource) handleQuery(client *azuredx.Client, q backend.DataQuery, user *backend.User) backend.DataResponse {
+func NewDatasource(settings backend.DataSourceInstanceSettings) (instancemgmt.Instance, error) {
+	datasourceSettings := models.DatasourceSettings{}
+	err := datasourceSettings.Load(settings)
+	if err != nil {
+		return nil, err
+	}
+
```

Comment:
You probable want to use `provider := httpclient.NewProvider(ProviderOptions)` here and add the auth token middleware through the provider options. Then you can put `provider` or `provider.New(settings.HTTPClientOptions())` as argument to `client.New`.

Reply:
:+1: